test/apps/iodemo/io_demo.cc

```diff
@@ -99,8 +139,14 @@ class P2pDemoCommon : public UcxContext {
     /* Asynchronous IO message */
     class IoMessage : public UcxCallback {
     public:
-        IoMessage(size_t buffer_size, io_op_t op, uint32_t sn, size_t data_size) :
-            _buffer(malloc(buffer_size)) {
+        IoMessage() :
+            _buffer(NULL) {
+        }
+        
+        void init(size_t buffer_size, io_op_t op, uint32_t sn, size_t data_size) {
+            if (_buffer == NULL) {
+                _buffer = malloc(buffer_size); 
```

Comment:
what if buffer_size is different then previous allocated size?
should pass buffer size to constructor and malloc in constructor

Reply:
> what if buffer_size is different then previous allocated size?
This depends on how we want to use this IoMessage class.  Normally we should put the items with the same structure to the same pool. Otherwise, should use different pools for them. If we want to use this class more complex, for example make the buffer size changeable., then we can compare the current size with previous size in init() function.  If the sizes are differen, then free the previous one, and allocate a new one. But I don't think we want to create this case in io demo, since this will cause performance drop. So I would say we should avoid allocate different buffer size for the items in the same pool. Please comment more if have different points. Thanks!
> should pass buffer size to constructor and malloc in constructor
Maybe not. If we allocate the buffer in constructor, then we should put the size of the buffer as parameter of constructor. The constructor function is excuted in MemoryPool::get(). We don't want to deal with the buffer size in MemoryPool class since this is not a concept of MemoryPool class. MemoryPool just allocate the items, but don't care the allocation of buffers(other resource) of items. Let me know please if any different points. Thanks!

Reply:
can we allocate the buffer in constructor?